optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/BiLeftHandSide.java

```diff
@@ -0,0 +1,287 @@
+        Variable<NewB> newB =
+                (Variable<NewB>) variableFactory.createVariable("newB", from(groupKey, k -> k.b));
+        return new BiLeftHandSide<>(new PatternVariable<>(newA, singletonList(groupByPattern)),
+                new PatternVariable<>(newB), variableFactory);
+    }
+
+    public <NewA, NewB> BiLeftHandSide<NewA, NewB> andGroupBy(BiFunction<A, B, NewA> keyMappingA,
+            BiConstraintCollector<A, B, ?, NewB> collectorB) {
+        Variable<A> inputA = patternVariableA.getPrimaryVariable();
+        Variable<B> inputB = patternVariableB.getPrimaryVariable();
+        Variable<BiTuple<A, B>> accumulateSource =
+                (Variable<BiTuple<A, B>>) variableFactory.createVariable(BiTuple.class, "source");
+        PatternVariable<B> newPatternVariableB = patternVariableB.bind(accumulateSource, inputA,
+                (b, a) -> new BiTuple<>(a, b));
+        Variable<NewA> groupKey = variableFactory.createVariable("groupKey");
+        Variable<NewB> output = variableFactory.createVariable("output");
+        ViewItem<?> innerGroupByPattern = joinViewItemsWithLogicalAnd(patternVariableA, newPatternVariableB);
+        ViewItem<?> groupByPattern = groupBy(innerGroupByPattern, inputA, inputB, groupKey,
+                keyMappingA::apply,
+                accFunction(() -> new DroolsBiAccumulateFunction<>(collectorB), accumulateSource).as(output));
+        Variable<NewA> newA = (Variable<NewA>) variableFactory.createVariable("newA", from(groupKey));
+        Variable<NewB> newB = (Variable<NewB>) variableFactory.createVariable("newB", from(output));
+        return new BiLeftHandSide<>(new PatternVariable<>(newA, singletonList(groupByPattern)),
+                new PatternVariable<>(newB), variableFactory);
+    }
+
+    public <NewA, NewB, NewC> TriLeftHandSide<NewA, NewB, NewC> andGroupBy(BiFunction<A, B, NewA> keyMappingA,
+            BiFunction<A, B, NewB> keyMappingB, BiConstraintCollector<A, B, ?, NewC> collectorC) {
+        Variable<A> inputA = patternVariableA.getPrimaryVariable();
+        Variable<B> inputB = patternVariableB.getPrimaryVariable();
+        Variable<BiTuple<A, B>> accumulateSource =
+                (Variable<BiTuple<A, B>>) variableFactory.createVariable(BiTuple.class, "source");
+        PatternVariable<B> newPatternVariableB = patternVariableB.bind(accumulateSource, inputA,
+                (b, a) -> new BiTuple<>(a, b));
+        Variable<BiTuple<NewA, NewB>> groupKey =
+                (Variable<BiTuple<NewA, NewB>>) variableFactory.createVariable(BiTuple.class, "groupKey");
+        Variable<NewC> output = variableFactory.createVariable("output");
+        ViewItem<?> innerGroupByPattern = joinViewItemsWithLogicalAnd(patternVariableA, newPatternVariableB);
+        ViewItem<?> groupByPattern = groupBy(innerGroupByPattern, inputA, inputB, groupKey,
+                (a, b) -> new BiTuple<>(keyMappingA.apply(a, b), keyMappingB.apply(a, b)),
```

Comment:
I think the trick with joining two columns into a `BiTuple` is worth documenting; mainly the motivation.